lib/lte_link_control/lte_lc.c

```diff
@@ -151,32 +182,328 @@ static const char cgauth[] = "AT+CGAUTH="CONFIG_LTE_PDN_AUTH;
 static const char legacy_pco[] = "AT%XEPCO=0";
 #endif
 
-void at_handler(void *context, const char *response)
+enum lte_lc_notif_type {
+	LTE_LC_NOTIF_CEREG,
+	LTE_LC_NOTIF_CSCON,
+	LTE_LC_NOTIF_CEDRXP,
+
+	LTE_LC_NOTIF_COUNT,
+};
+
+static const char *const at_notifs[] = {
+	[LTE_LC_NOTIF_CEREG] = "+CEREG",
+	[LTE_LC_NOTIF_CSCON] = "+CSCON",
+	[LTE_LC_NOTIF_CEDRXP] = "+CEDRXP",
+};
+
+BUILD_ASSERT(ARRAY_SIZE(at_notifs) == LTE_LC_NOTIF_COUNT);
+
+static bool is_relevant_notif(const char *notif, enum lte_lc_notif_type *type)
+{
+	for (size_t i = 0; i < ARRAY_SIZE(at_notifs); i++) {
+		if (strncmp(at_notifs[i], notif,
+			    strlen(at_notifs[i])) == 0) {
+			/* The notification type matches the array index */
+			*type = i;
+
+			return true;
+		}
+	}
+
+	return false;
+}
+
+static int parse_cereg(const char *notification,
+		       enum lte_lc_nw_reg_status *reg_status,
+		       struct lte_lc_cell *cell,
+		       struct lte_lc_psm_cfg *psm_cfg)
+{
+	int err, status;
+	struct at_param_list resp_list = {0};
+	char str_buf[10];
+	size_t len = sizeof(str_buf) - 1;
+
+	if (notification == NULL) {
+		return -EINVAL;
+	}
+
+	err = at_params_list_init(&resp_list, AT_CEREG_PARAMS_COUNT_MAX);
+	if (err) {
+		LOG_ERR("Could not init AT params list, error: %d", err);
+		return err;
+	}
+
+	/* Parse CEREG response and populate AT parameter list */
+	err = at_parser_max_params_from_str(notification,
+					    NULL,
+					    &resp_list,
+					    AT_CEREG_PARAMS_COUNT_MAX);
+	if (err) {
+		LOG_ERR("Could not parse AT+CEREG response, error: %d", err);
+		goto clean_exit;
+	}
+
+	/* Parse network registration status */
+	err = at_params_int_get(&resp_list,
+				AT_CEREG_REG_STATUS_INDEX,
+				&status);
+	if (err) {
+		LOG_ERR("Could not get registration status, error: %d", err);
+		goto clean_exit;
+	}
+
+	*reg_status = status;
+
+	/* Parse tracking area code */
+	err = at_params_string_get(&resp_list,
+				   AT_CEREG_TAC_INDEX,
+				   str_buf, &len);
+	if (err) {
+		LOG_ERR("Could not get tracking area code, error: %d", err);
+		goto clean_exit;
+	}
+
+	str_buf[len] = '\0';
+	cell->tac = strtoul(str_buf, NULL, 16);
+
+	/* Parse cell ID */
+	len = sizeof(str_buf) - 1;
+
+	err = at_params_string_get(&resp_list,
+				   AT_CEREG_CELL_ID_INDEX,
+				   str_buf, &len);
+	if (err) {
+		LOG_ERR("Could not get cell ID, error: %d", err);
+		goto clean_exit;
+	}
+
+	str_buf[len] = '\0';
+	cell->id = strtoul(str_buf, NULL, 16);
+
+	/* Parse PSM configuration */
+	err = parse_psm_cfg(&resp_list, true, psm_cfg);
+	if (err) {
+		LOG_ERR("Failed to parse PSM configuration, error: %d", err);
+		goto clean_exit;
+	}
+
+clean_exit:
+	at_params_list_free(&resp_list);
+
+	return err;
+}
+
+static void at_handler(void *context, const char *response)
 {
 	ARG_UNUSED(context);
 
 	int err;
-	enum lte_lc_nw_reg_status status;
+	bool notify = false;
+	enum lte_lc_notif_type notif_type;
+	struct lte_lc_evt evt;
 
 	if (response == NULL) {
```

Comment:
Is this not already done in the `is_relevant_notif` function?

Reply:
Not quite sure what you refer to here. The NULL check? It's only done here, before calling `is_relevant_notif()` on the response. Can be argued we could drop the check altogether though

Reply:
You should not check twice if `response` is NULL but you can keep it for clarity.
If the pointer in `strncmp` is NULL, you will ignore the notification anyway.